### PR TITLE
Attempts at making the GL backend cache more state

### PR DESCRIPTION
**Pull request by icefoxen**
Would resolve (a large part of) issue #1198 .

It's quite a brute-force solution, and I am deeply uncertain that it tracks everything correctly, but I tried to err on the side of safety, allowing redundant state-setting calls rather than getting rid of them.  It is also incomplete: there is some stuff, such as stencil buffers, that it doesn't try to track (because I don't really know how).  However, it appears to work; I don't *see* any differences in any of the examples, and apitrace output for the performance example is about equivalent to the gl example (which is good!).

Still, please sanity-check this, and help make sure it works with fancy stuff like multiple PSO and vertex type definitions.  I'll be doing the same if/when I have time.

### CODE REVIEW

**Comment by icefoxen** [comment]
Dammit, hang on...

**Comment by msiglreith** [comment]
Looks like your branch is based on the `master` branch but the PR is opened for merging into `0.14`

**Comment by Bastacyclop** [comment]
@icefoxen It seems that your commits are all before the "random" ones, maybe you could try to do `git reset   67ee0eb` and see if `git log` is ok then ?

**Comment by icefoxen** [comment]
@msiglreith It was SUPPOSED to be based on 0.14... let me try again.  x_X

**Comment by icefoxen** [comment]
Closed in favor of #1221
